From a0f9b13ebdd1450b6a51b27e2a5d3762bc2f64d3 Mon Sep 17 00:00:00 2001
From: Marcin Wojtas <mw@semihalf.com>
Date: Thu, 20 Dec 2018 10:27:21 +0100
Subject: [PATCH 0828/1051] net: mvpp2: force TX FIFO transactions priority on
 the AXI bus

It may happen that when there are processes which load
the internal AXI bus with CPU-DRAM transactions, it is
possible to observe the TX FIFO underrun errors. It means,
that the PP2 port is unable to fetch the data from DRAM
on time during egress.

In order to avoid such issue, prioritize TX FIFO AXI bus
access using the PP2 NIC AXI bridge configuration.

Change-Id: I3d36776e911396e20254a319c3e8f87d76a23e61
Signed-off-by: Marcin Wojtas <mw@semihalf.com>
Reviewed-on: https://sj1git1.cavium.com/1177
Tested-by: sa_ip-sw-jenkins
Reviewed-by: Stefan Chulski <Stefan.Chulski@cavium.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2/mvpp2.h      | 1 +
 drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c | 4 +++-
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2/mvpp2.h b/drivers/net/ethernet/marvell/mvpp2/mvpp2.h
index e4b4b38ca011..84d7bb0ae047 100644
--- a/drivers/net/ethernet/marvell/mvpp2/mvpp2.h
+++ b/drivers/net/ethernet/marvell/mvpp2/mvpp2.h
@@ -223,6 +223,7 @@
 #define MVPP22_AXI_RXQ_DESCR_WR_ATTR_REG	0x411c
 #define MVPP22_AXI_RX_DATA_WR_ATTR_REG		0x4120
 #define MVPP22_AXI_TX_DATA_RD_ATTR_REG		0x4130
+#define     MVPP22_AXI_TX_DATA_RD_QOS_ATTRIBUTE (0x3 << 4)
 #define MVPP22_AXI_RD_NORMAL_CODE_REG		0x4150
 #define MVPP22_AXI_RD_SNOOP_CODE_REG		0x4154
 #define MVPP22_AXI_WR_NORMAL_CODE_REG		0x4160
diff --git a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
index 5c3a314deb20..e93ba5468318 100644
--- a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
@@ -6117,7 +6117,9 @@ static void mvpp2_axi_init(struct mvpp2 *priv)
 	mvpp2_write(priv, MVPP22_AXI_RXQ_DESCR_WR_ATTR_REG, wrval);
 
 	/* Buffer Data */
-	mvpp2_write(priv, MVPP22_AXI_TX_DATA_RD_ATTR_REG, rdval);
+	/* Force TX FIFO transactions priority on the AXI QOS bus */
+	mvpp2_write(priv, MVPP22_AXI_TX_DATA_RD_ATTR_REG,
+		    rdval | MVPP22_AXI_TX_DATA_RD_QOS_ATTRIBUTE);
 	mvpp2_write(priv, MVPP22_AXI_RX_DATA_WR_ATTR_REG, wrval);
 
 	val = MVPP22_AXI_CODE_CACHE_NON_CACHE
-- 
2.17.1

